File: Week 6/cpp/4.prefix_sum_and_kadane.cpp
/*
 * WEEK 6 - C++ DSA
 * Topic: Prefix Sum & Kadane's Algorithm
 * File: 4.prefix_sum_and_kadane.cpp
 *
 * CONCEPT:
 *     Prefix sum: build P[i] = arr[0]+...+arr[i] in O(n); answer any
 *     range sum [l,r] in O(1) as P[r] - (l ? P[l-1] : 0).
 *     Kadane: maximum contiguous subarray sum in O(n) via the DP
 *         current = max(arr[i], current + arr[i])
 *         best    = max(best, current)
 *
 * KEY POINTS:
 *     - Always seed Kadane with arr[0] to handle all-negative arrays.
 *     - Extended Kadane tracks (start, end) indices.
 *     - std::partial_sum (header <numeric>) is the STL prefix-sum.
 *
 * ALGORITHM / APPROACH:
 *     buildPrefix:
 *         P[0] = arr[0]; P[i] = P[i-1] + arr[i]
 *     rangeSum(l,r):
 *         (l == 0) ? P[r] : P[r] - P[l-1]
 *     kadane:
 *         best = current = arr[0]
 *         for i in [1..n): current = max(arr[i], current+arr[i]);
 *                          best    = max(best, current)
 *         return best
 *
 * C++-SPECIFIC NOTES:
 *     - Use long long for prefix sums to avoid overflow on large n.
 *     - std::max is in <algorithm>.
 *     - Tuple/array returns can use std::tuple or struct.
 *
 * DRY RUN:
 *     arr = {3,-1,2,4,-3,7}; P = {3,2,4,8,5,12}
 *     rangeSum(1,4) = 5 - 3 = 2
 *
 *     Kadane on {-2,1,-3,4,-1,2,1,-5,4}:
 *         current/best track to (6, [4,-1,2,1])
 *
 * COMPLEXITY:
 *     Prefix : O(n) build, O(1) per query, O(n) space
 *     Kadane : O(n) time, O(1) space
 */

#include <iostream>
#include <vector>
#include <algorithm>
#include <numeric>
#include <tuple>

std::vector<long long> buildPrefix(const std::vector<int>& arr) {
    std::size_t n = arr.size();
    std::vector<long long> prefix(n);
    if (n == 0) return prefix;
    prefix[0] = arr[0];
    for (std::size_t i = 1; i < n; ++i) prefix[i] = prefix[i - 1] + arr[i];
    return prefix;
}

long long rangeSum(const std::vector<long long>& prefix, int l, int r) {
    if (l == 0) return prefix[r];
    return prefix[r] - prefix[l - 1];
}

int maxSubarraySum(const std::vector<int>& arr) {
    int best = arr[0], current = arr[0];
    for (std::size_t i = 1; i < arr.size(); ++i) {
        current = std::max(arr[i], current + arr[i]);
        best = std::max(best, current);
    }
    return best;
}

std::tuple<int,int,int> maxSubarrayWithIndices(const std::vector<int>& arr) {
    int best = arr[0], current = arr[0];
    int start = 0, end = 0, tempStart = 0;
    for (std::size_t i = 1; i < arr.size(); ++i) {
        if (arr[i] > current + arr[i]) {
            current = arr[i];
            tempStart = static_cast<int>(i);
        } else {
            current += arr[i];
        }
        if (current > best) {
            best = current;
            start = tempStart;
            end = static_cast<int>(i);
        }
    }
    return {best, start, end};
}

void printVec(const std::vector<int>& arr) {
    std::cout << "[";
    for (std::size_t i = 0; i < arr.size(); ++i)
        std::cout << arr[i] << (i + 1 < arr.size() ? ", " : "");
    std::cout << "]";
}

void printVec(const std::vector<long long>& arr) {
    std::cout << "[";
    for (std::size_t i = 0; i < arr.size(); ++i)
        std::cout << arr[i] << (i + 1 < arr.size() ? ", " : "");
    std::cout << "]";
}

int main() {
    std::vector<int> arr = {3, -1, 2, 4, -3, 7};
    auto prefix = buildPrefix(arr);
    std::cout << "Array:  "; printVec(arr);    std::cout << "\n";
    std::cout << "Prefix: "; printVec(prefix); std::cout << "\n";
    std::cout << "Sum of arr[1..4] = " << rangeSum(prefix, 1, 4) << "\n";
    std::cout << "Sum of arr[0..5] = " << rangeSum(prefix, 0, 5) << "\n";
    std::cout << "Sum of arr[2..2] = " << rangeSum(prefix, 2, 2) << "\n";

    std::vector<int> test1 = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
    std::cout << "\nArray: "; printVec(test1); std::cout << "\n";
    std::cout << "Max subarray sum: " << maxSubarraySum(test1) << "\n";
    auto [best, s, e] = maxSubarrayWithIndices(test1);
    std::cout << "Max sum: " << best << " | Subarray arr[" << s << ".." << e << "] = [";
    for (int i = s; i <= e; ++i) std::cout << test1[i] << (i < e ? ", " : "");
    std::cout << "]\n";

    std::vector<int> allNeg = {-5, -1, -8, -3};
    std::cout << "\nAll-negative: "; printVec(allNeg); std::cout << "\n";
    std::cout << "Max sum: " << maxSubarraySum(allNeg) << "\n";

    std::vector<int> single = {42};
    std::cout << "\nSingle element [42]: max sum = "
              << maxSubarraySum(single) << "\n";

    return 0;
}

/*
 * NOTES — C++ vs Java:
 *     - std::partial_sum can build the prefix array in one call.
 *     - std::max takes any number of args via initializer_list overload.
 *     - long long is safer for accumulations than int.
 *     - Structured bindings make tuple returns ergonomic.
 */

File: Week 6/rust/4.prefix_sum_and_kadane.rs
/*
 * WEEK 6 - RUST DSA
 * Topic: Prefix Sum & Kadane's Algorithm
 * File: 4.prefix_sum_and_kadane.rs
 *
 * CONCEPT:
 *     A) Prefix sum: P[i] = arr[0]+...+arr[i]; range sum is O(1).
 *     B) Kadane: maximum contiguous subarray sum in O(n).
 *
 * KEY POINTS:
 *     - Use i64 for prefix sums; reduces overflow risk.
 *     - Kadane handles all-negative input by seeding with arr[0].
 *     - Extended Kadane returns (sum, start, end).
 *
 * ALGORITHM / APPROACH:
 *     Build prefix:
 *         prefix[0] = arr[0]; prefix[i] = prefix[i-1] + arr[i]
 *     range_sum(l,r):
 *         (l == 0) ? prefix[r] : prefix[r] - prefix[l-1]
 *     Kadane:
 *         best = current = arr[0]
 *         for i in 1..n:
 *             current = max(arr[i], current + arr[i])
 *             best    = max(best, current)
 *
 * RUST-SPECIFIC NOTES:
 *     - i32::max(a, b) is the in-prelude max for two i32 values.
 *     - Iterators: arr.iter().scan(0, |acc, &x| { *acc += x; Some(*acc) })
 *       lazily yields a prefix-sum iterator.
 *
 * DRY RUN:
 *     arr = [3,-1,2,4,-3,7]
 *     P   = [3, 2,4,8, 5,12]
 *     range_sum(1,4) = 5 - 3 = 2
 *
 *     Kadane on [-2,1,-3,4,-1,2,1,-5,4] -> 6 from [4,-1,2,1].
 *
 * COMPLEXITY:
 *     Prefix : O(n) build, O(1) per query
 *     Kadane : O(n) time, O(1) space
 */

fn build_prefix(arr: &[i64]) -> Vec<i64> {
    let n = arr.len();
    let mut prefix = vec![0_i64; n];
    if n == 0 {
        return prefix;
    }
    prefix[0] = arr[0];
    for i in 1..n {
        prefix[i] = prefix[i - 1] + arr[i];
    }
    prefix
}

fn range_sum(prefix: &[i64], l: usize, r: usize) -> i64 {
    if l == 0 {
        prefix[r]
    } else {
        prefix[r] - prefix[l - 1]
    }
}

fn max_subarray_sum(arr: &[i64]) -> i64 {
    let mut best = arr[0];
    let mut current = arr[0];
    for i in 1..arr.len() {
        current = std::cmp::max(arr[i], current + arr[i]);
        best = std::cmp::max(best, current);
    }
    best
}

fn max_subarray_with_indices(arr: &[i64]) -> (i64, usize, usize) {
    let mut best = arr[0];
    let mut current = arr[0];
    let mut start: usize = 0;
    let mut end: usize = 0;
    let mut temp_start: usize = 0;

    for i in 1..arr.len() {
        if arr[i] > current + arr[i] {
            current = arr[i];
            temp_start = i;
        } else {
            current += arr[i];
        }
        if current > best {
            best = current;
            start = temp_start;
            end = i;
        }
    }
    (best, start, end)
}

fn main() {
    let arr: Vec<i64> = vec![3, -1, 2, 4, -3, 7];
    let prefix = build_prefix(&arr);
    println!("Array:  {:?}", arr);
    println!("Prefix: {:?}", prefix);
    println!("Sum of arr[1..4] = {}", range_sum(&prefix, 1, 4));
    println!("Sum of arr[0..5] = {}", range_sum(&prefix, 0, 5));
    println!("Sum of arr[2..2] = {}", range_sum(&prefix, 2, 2));

    let test1: Vec<i64> = vec![-2, 1, -3, 4, -1, 2, 1, -5, 4];
    println!("\nArray: {:?}", test1);
    println!("Max subarray sum: {}", max_subarray_sum(&test1));
    let (best, s, e) = max_subarray_with_indices(&test1);
    println!(
        "Max sum: {} | Subarray arr[{}..{}] = {:?}",
        best,
        s,
        e,
        &test1[s..=e]
    );

    let all_neg: Vec<i64> = vec![-5, -1, -8, -3];
    println!("\nAll-negative: {:?}", all_neg);
    println!("Max sum: {}", max_subarray_sum(&all_neg));

    let single: Vec<i64> = vec![42];
    println!("\nSingle element {:?}: max sum = {}", single, max_subarray_sum(&single));
}

/*
 * NOTES — Rust vs Java:
 *     - i64 instead of int gives a comfortable overflow margin.
 *     - Tuple returns (i64, usize, usize) replace the Java int[] trick.
 *     - Slicing &test1[s..=e] is range-inclusive and zero-copy.
 *     - std::cmp::max works for any Ord type; built into the prelude.
 */
